### ISSUE TITLE
Unexpected .load() behaviour when nodata is not set 

### ISSUE THREAD

**Issue by Alex-Mackie** [opendatacube/odc-stac#162]
We have some STAC assets that don't have nodata values (every coordinate contains a valid value) and therefore the `nodata` property of the raster extension is not set. 

The problem is that when users load this data with `.load()` they'll get a partial Dataset without warning. 

Compare:

`odc_stac.load(search.items(), chunks={'x':8000, 'y':8000}, resolution=0.1).FC.plot()`

![image](https://github.com/opendatacube/odc-stac/assets/49118224/2e358d46-4be1-48c6-8b69-69c1f737a71d)

with

`odc_stac.load(search.items(), chunks={'x':8000, 'y':8000}, nodata=999, resolution=0.1).FC.plot()`

![image](https://github.com/opendatacube/odc-stac/assets/49118224/5099602b-bec8-467d-bf85-e97d14aa6988)

We'd like to protect our users from this. It isn't intuitive to all users that combining/merging arrays in xArray will often require a safe nodata value to be supplied so they'd typically be mystified.

Should we update our catalogue with a nodata value that would be "safe" to use in these circumstances (even though it never actually occurs) or should `odc_stac.load()` raise an error or warning in a situation like this where merging/combining is implied but no nodata value is supplied (either via the STAC metadata or as an argument to `.load()`)?

**Comment by Kirill888** [comment]
@Alex-Mackie what is the data type of your pixels? If it's `float32` then this could be an error in `odc-stac` as it should default to using `nan` as a fill value and not `0`. If it's not `float32` but something like `int16` then this still looks like an error in odc-stac to me as it should be treating right-side of an image as "empty" and paste the other dataset there without an issue, but clearly it treats that part as "filled".

`nodata` has two meanings really:
1. a replacement for `NaN` when using non-float data types, a cheaper version of a separate mask image
2. fill value, default value for pixels outside of the data coverage

To configure `nodata` in your stac catalog refer to this:

https://odc-stac.readthedocs.io/en/latest/stac-best-practice.html

To summarize: what you are reporting is probably an error. You can work around the error by explicitly forcing a specific "fill value" outside of the valid range of your data. If your data is floating point, try forcing `nodata=float("nan")` and see if that fixes it.

**Comment by Alex-Mackie** [comment]
Thanks Kirill, we get this behaviour on uint8 & uint16 assets.

I'm beginning to think it is prudent to supply a nodata value anyway in these situations (as a general STAC-publishing practise) to function as a prompt for what to use outside the data coverage. The only time this would be tricky is if you are using the whole range of your dtype (which is not typical in our use cases).

**Comment by Kirill888** [comment]
Even if your source data has no holes in it, it is still useful to allocate a value to mean nan if possible. It’s harder with single byte rasters. This is usually handled with mask arrays stored next to pixel values, unfortunately odc-stac currently doesn’t support data masked in that way properly.

Most tested case is probably int16 with nodata value set.

**Comment by Kirill888** [comment]
@Alex-Mackie do you mind testing your failing case with the latest versions of odc-stac and odc-geo? There were a bit of rework around that area.

**Comment by Alex-Mackie** [comment]
I was using
```
odc-geo==0.4.6
odc-stac==0.3.9
```
I can confirm I see the same behaviour after updating to:
```
odc-geo==0.4.7
odc-stac==0.3.10
```

**Comment by Alex-Mackie** [comment]
I haven't dug any deeper into the where this is coming from but I can recreate somewhat similar behaviour with publicly available MODIS Burned Area data from the Planetary Computer Catalogue:

```
# Get Modis burned area for northern South America

collection_name = "modis-64A1-061"
start_date = "2010-01-01"
end_date = "2010-01-31"
time_range = f"{start_date}/{end_date}"

catalog = pystac_client.Client.open(
    "https://planetarycomputer.microsoft.com/api/stac/v1",
    modifier=planetary_computer.sign_inplace,
)

aoi = {
  "type": "Polygon",
  "coordinates": [
    [
      [-86.36253169971852, -6.597051557427108],
      [-46.85884169908314, -6.597051557427108],
      [-46.85884169908314, 16.056895243448793],
      [-86.36253169971852, 16.056895243448793],
      [-86.36253169971852, -6.597051557427108]
    ]
  ]
}

search = catalog.search(filter_lang="cql2-json", filter={
  "op": "and",
  "args": [
    {"op": "s_intersects", "args": [{"property": "geometry"}, aoi]},
    {"op": "=", "args": [{"property": "collection"}, collection_name]},
    {"op": "t_intersects", "args": [{"property": "datetime"}, time_range]},
  ]
})

items = search.item_collection()
```

## 1. Native projection with no nodata value supplied, FAIL
We have "not burned" out in the ocean
```
ds = odc.stac.load(
    items,
    bands=["Burn_Date"],
)
```
![Image](https://github.com/user-attachments/assets/4123c83c-4dd5-4f53-aaf7-bc819619477e)

## 2, Different projection with no nodata value supplied, FAIL
We have the same problem as 1. and also a lot of interesting nodata areas

```
ds = odc.stac.load(
    items,
    bands=["Burn_Date"],
    crs="EPSG:6933",
)
```
![Image](https://github.com/user-attachments/assets/6048b5a1-1f45-4879-9a7f-f785fed0c713)

## 3. Native projection with nodata value supplied - PASS
```
ds = odc.stac.load(
    items,
    bands=["Burn_Date"],
    nodata=-1
)
```
![Image](https://github.com/user-attachments/assets/200d30d2-409f-4728-99e8-ea03d54282e1)

## 4. Different projection with nodata value supplied, PASS
```
ds = odc.stac.load(
    items,
    bands=["Burn_Date"],
    crs="EPSG:6933",
    nodata=-1
)
```
![Image](https://github.com/user-attachments/assets/89d1ec6d-68f8-4964-af17-8d0052c2dc5d)

**Comment by Kirill888** [comment]
@Alex-Mackie thank you for this investigation, but this is not a surprise to me. 

1. Items do not include `raster:nodata` field
2. User doesn't specify `nodata=`
3. `odc.stac.load` still needs to pick `fill_value` so it chooses `0` as a default
4. `0` clashes with valid pixel values of `Unburned` class: https://lpdaac.usgs.gov/products/mcd64a1v006/
5. Setting `nodata=-1` fixes the problem

In conclusion when `nodata` is not set in the metadata AND is not set during `odc.stac.load` AND pixel type is not float,  then fill value of `0` will be picked. When `0` clashes with a valid pixel value/class results will look bad.

Green blob in the middle of the sea in the first example is not covered by any asset footprint, so it's left with default `fill_value` of `0`. There are no issues along the seams since MODIS tiles do not overlap, so they are just pasted next to each and `nodata` is of no relevance. When you change projection, non-overlapping tiles become overlapping and now fusing needs to happen, since `nodata` is not configured all pixels of projected image are considered as valid and we pick first valid observed pixel as output. This results in black stripes. And this is why collections like Sentinel-2 and Landsat have overlapping "tiles" and a well defined `nodata` field.

**Comment by Kirill888** [comment]
consider raising an issue with planetary computer folks to configure `nodata` in the metadata.
